File: library/src/main/java/io/blackbox_vision/mvphelpers/logic/presenter/BasePresenter.java
package io.blackbox_vision.mvphelpers.logic.presenter;

import android.support.annotation.NonNull;

import io.blackbox_vision.mvphelpers.logic.view.BaseView;


public abstract class BasePresenter<V extends BaseView> {
    private V view;

    protected abstract void onViewAttached(@NonNull V view);

    protected abstract void onViewDetached();

    public boolean isViewAttached() {
        return this.view != null;
    }

    public void attachView(@NonNull V view) {
        this.view = view;
        onViewAttached(view);
    }

    public void detachView() {
        this.view = null;
        onViewDetached();
    }

    public V getView() {
        return view;
    }
}

File: library/src/main/java/io/blackbox_vision/mvphelpers/logic/presenter/BasePresenter.kt
package io.blackbox_vision.mvphelpers.logic.presenter

import io.blackbox_vision.mvphelpers.logic.view.BaseView


abstract class BasePresenter<T : BaseView> {
    var view: T? = null
        private set

    protected val isViewAttached: Boolean
        get() = this.view != null

    abstract fun onViewAttached(view: T?)

    abstract fun onViewDetached(view: T?)

    fun attachView(view: T) {
        this.view = view
        onViewAttached(view)
    }

    fun detachView() {
        onViewDetached(this.view)
        this.view = null
    }
}
